### PR TITLE
Endpoint to export artifact version status per environment

### PR DESCRIPTION
**Pull request by robfletcher**
As discussed in #643.

### CODE REVIEW

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt [hunk 1]
```diff
@@ -167,10 +179,26 @@ abstract class ArtifactRepositoryTests<T : ArtifactRepository> : JUnit5Minutests
...
           }
         }
+
+        test("the artifact version is pending in the environment") {
+          val environment1Versions = subject
+            .versionsByEnvironment(manifest)
+            .first { it.name == environment1.name }
+            .artifacts
+            .first { it.name == artifact1.name && it.type == artifact1.type }
+            .versions
```
Comment (luispollo) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt]: These lines repeat across all new tests... Would be nice to extract them into a function, maybe?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/memory/InMemoryArtifactRepository.kt [hunk 2]
```diff
@@ -138,6 +146,36 @@ class InMemoryArtifactRepository : ArtifactRepository {
     }
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> =
+    deliveryConfig
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/memory/InMemoryArtifactRepository.kt]: I find this hard to parse, with the super nested structure. Any way we can simplify to make it easier to understand what's going on?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/EnvironmentArtifactsSummary.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+data class ArtifactVersions(
+  val name: String,
+  val type: ArtifactType,
+  val versions: Map<PromotionStatus, List<String>>
+)
+
+enum class PromotionStatus {
+  PREVIOUS, CURRENT, DEPLOYING, PENDING
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/EnvironmentArtifactsSummary.kt]: I think you might need another state here. 

There are potentially many artifacts that were approved for an environment but not deployed to it. How are you representing that state?

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 4]
```diff
@@ -215,6 +225,113 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
+    return deliveryConfig.environments.map { environment ->
+      val artifactVersions = deliveryConfig.artifacts.map { artifact ->
+        //          select e.name, eav.artifact_version, eav.approved_at, eav.deployed_at
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Remove this code?

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EnvironmentController.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping(path = ["/environments"])
+class EnvironmentController(
+  private val deliveryConfigRepository: DeliveryConfigRepository
+) {
+
+  @GetMapping(
```
Comment (emjburns) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EnvironmentController.kt]: do you need or want a similar api at the delivery config level?

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 6]
```diff
@@ -215,6 +225,113 @@ class SqlArtifactRepository(
...
+//          and da.name = 'keel'
+//          and da.type = 'DEB'
+//          and e.delivery_config_uid = dc.uid
+//          and dc.name = 'keel-manifest'
+//          and not exists (
+//          select 1 from environment_artifact_versions eav
+//          where eav.artifact_version = dav.version
+//          and eav.environment_uid = e.uid
+//        )
+        val rawResults = jooq
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: This query is so large and confusing. Could you put a function description that describes the steps you're taking to build this information?

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 7]
```diff
@@ -215,6 +225,113 @@ class SqlArtifactRepository(
...
+              .and(ENVIRONMENT.NAME.eq(environment.name))
+              .andNotExists(
+                selectOne()
+                  .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+                  .where(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(DELIVERY_ARTIFACT_VERSION.VERSION))
+                  .and(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+              )
+          )
+          .fetch()
+        val turnt = rawResults.map { (version, approvedAt, deployedAt) ->
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: what does `turnt` mean?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/EnvironmentArtifactsSummary.kt (hunk 3 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/EnvironmentArtifactsSummary.kt]: Yeah, right now we're not and they show up as `DEPLOYING` which is a problem I only noticed when I pointed it at real data.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 4 above)
Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: D'oh

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 7 above)
Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: It means I forgot to inline some variables

**Comment by robfletcher** [comment]
I think the query will likely change. We may even want to explicitly store data to more concretely back this API. The important thing right now is that it behaves correctly.

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt (hunk 1 above)
Comment (robfletcher) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt]: I made it an extension on the fixture

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 8]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
...
+            deploying = deploying?.version,
+            pending = pending.map { it.version },
+            previous = previous.map { it.version }
+          )
+        )
+      }
+      EnvironmentArtifactsSummary(environment.name, artifactVersions)
+    }
+  }
+
```
Comment (robfletcher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Hopefully more readable now

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 9]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
```
Comment (asher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Can we change the method signature to support limiting? For `versionsInEnvironment`, it would probably be ok to `order by ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT desc` and take the topN.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 10]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
...
+        val deploying = versions.filter { it.approvedAt != null && it.deployedAt == null }.maxBy { checkNotNull(it.approvedAt) }
+        val pending = versions.filter { it.approvedAt == null }
+        ArtifactVersions(
+          name = artifact.name,
+          type = artifact.type,
+          versions = ArtifactVersionStatus(
+            current = current?.version,
+            deploying = deploying?.version,
+            pending = pending.map { it.version }.sortedWith(artifact.versioningStrategy.comparator),
+            previous = previous.map { it.version }.sortedWith(artifact.versioningStrategy.comparator)
```
Comment (robfletcher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: @emjburns good job with that `versioningStrategy`. I was worried sorting this stuff was going to be a PITA.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 9 above)
Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Could we just default to a limit? I worry that it seems misleading to specify a limit in the signature that we may blow because it's the product of multiple queries. I don't think there's any reliable way to sort the 2nd query in SQL.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 11]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
+    return deliveryConfig.environments.map { environment ->
```
Comment (asher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: It looks like the response is entirely built from the `ENVIRONMENT_ARTIFACT_VERSIONS` table, so another approach could be to: 

1) get all envIds for the delivery config
2) select * from ENVIRONMENT_ARTIFACT_VERSIONS where ENVIRONMENT_UID in (....) and I think `versionsInEnvironment` and `pendingVersions` can both be built from that. I've found having jooq fetch into a `ResultSet` is helpful when deriving disparate things from a single query. 

I think it would still make sense to have a limit parameter on the method in this case, even though it would have to be implemented in memory, if only to limit the cost to browsers parsing the response.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 12]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
...
+          )
+        val versions = versionsInEnvironment
+          .unionAll(pendingVersions)
+          .fetch()
+          .map { (version, approvedAt, deployedAt) ->
+            ArtifactVersionResult(version, approvedAt, deployedAt)
+          }
+        val deployed = versions.filter { it.deployedAt != null }
+        val current = deployed.maxBy { checkNotNull(it.deployedAt) }
+        val previous = deployed.filterNot { it == current }
```
Comment (asher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: I don't think this is going to be right. If `deployed` was sorted by `deployedAt` in descending order, then we can `val previous = deployed.first { it != current }` but as written it looks like it might be an arbitrary assignment.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 11 above)
Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: It has to union that with a read from `DELIVERY_ARTIFACT_VERSION` where `NOT EXISTS` in `ENVIRONMENT_ARTIFACT_VERSIONS` because versions only appear on that latter table once they get approved.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 12 above)
Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: shit yes. Fixing

Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Urgh, why does Kotlin not have a combination of `sortedBy` and `sortedWith`. This is going to suck.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 13]
```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
...
+            DELIVERY_CONFIG
+          )
+          .where(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID))
+          .and(DELIVERY_ARTIFACT.NAME.eq(artifact.name))
+          .and(DELIVERY_ARTIFACT.TYPE.eq(artifact.type.name))
+          .and(ENVIRONMENT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID))
+          .and(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(DELIVERY_CONFIG.UID))
+          .and(DELIVERY_CONFIG.NAME.eq(deliveryConfig.name))
+          .and(ENVIRONMENT.NAME.eq(environment.name))
+        val pendingVersions = jooq
```
Comment (asher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: This needs to take artifact status into account since environments can be concerned with different sets of statuses within a delivery config. The current implementation would list `CANDIDATE` artifacts as pending on an environment that only accepts `RELEASE`.

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EnvironmentController.kt (hunk 5 above)
Comment (robfletcher) [reply keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EnvironmentController.kt]: I don't know. Do you @erikmunson?

Comment (erikmunson) [reply keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EnvironmentController.kt]: I will 100% need one later when we move into the multi-app world, but for now in this moment I really don't know how to make use of that properly in deck so this is sufficient. There are so many fundamental assumptions we have about the app you're in _only_ ever showing things from the one app that I'm going to avoid figuring it out until we know more about what our desired UX is.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt (hunk 10 above)
Comment (emjburns) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Thanks!

**Comment by robfletcher** [comment]
@asher 2aad114 contains filtering by artifact status. I don't see a sane way to do it in the query. Trying to rustle together a test as well (I know, I know)

**Comment by robfletcher** [comment]
Ok dabe1c1 contains filtering by status with test coverage
